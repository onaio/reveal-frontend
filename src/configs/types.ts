import {
  adminLayerColors,
  GREEN_THRESHOLD,
  JurisdictionLevels,
  ORANGE_THRESHOLD,
  YELLOW_THRESHOLD,
} from './settings';

// threshold types
export type GREEN_THRESHOLD = typeof GREEN_THRESHOLD;
export type YELLOW_THRESHOLD = typeof YELLOW_THRESHOLD;
export type ORANGE_THRESHOLD = typeof ORANGE_THRESHOLD;

// admin layer color type
export type adminLayerColorsType = typeof adminLayerColors[number];

// jurisdiction types
export type JurisdictionTypes = typeof JurisdictionLevels[number];

/** ISO 3166-alpha-2 admin codes */
export type ADMN0_PCODE =
  | 'TH'
  | 'ZM'
  | 'ra Zambia'
  | 'NA'
  | 'BW'
  | 'Chadiza'
  | 'Sinda'
  | 'Katete'
  | 'Siavonga'
  | 'Lop Buri'
  | 'Oddar Meanchey Province'
  | 'Lusaka';

// tslint:disable-next-line:no-empty-interface
export interface EmptyObject {}
