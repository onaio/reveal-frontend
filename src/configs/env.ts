/** This module handles settings taken from environment variables */

/** The website name */
export const WEBSITE_NAME = process.env.REACT_APP_WEBSITE_NAME || 'Reveal';
export type WEBSITE_NAME = typeof WEBSITE_NAME;

/** The domain name */
export const DOMAIN_NAME = process.env.REACT_APP_DOMAIN_NAME || 'http://localhost:3000';
export type DOMAIN_NAME = typeof DOMAIN_NAME;

/** Do you want to enable the IRS features? */
export const ENABLE_IRS = process.env.REACT_APP_ENABLE_IRS === 'true';
export type ENABLE_IRS = typeof ENABLE_IRS;

/** Do you want to enable the Focus Investigation features? */
export const ENABLE_FI = process.env.REACT_APP_ENABLE_FI === 'true';
export type ENABLE_FI = typeof ENABLE_FI;

/** Do you want to enable the users page? */
export const ENABLE_USERS = process.env.REACT_APP_ENABLE_USERS === 'true';
export type ENABLE_USERS = typeof ENABLE_USERS;

/** Do you want to enable the about page? */
export const ENABLE_ABOUT = process.env.REACT_APP_ENABLE_ABOUT === 'true';
export type ENABLE_ABOUT = typeof ENABLE_ABOUT;

/** Do you want to disable login protection? */
export const DISABLE_LOGIN_PROTECTION = process.env.REACT_APP_DISABLE_LOGIN_PROTECTION === 'true';
export type DISABLE_LOGIN_PROTECTION = typeof DISABLE_LOGIN_PROTECTION;

/** The Superset API base */
export const SUPERSET_API_BASE = process.env.REACT_APP_SUPERSET_API_BASE || 'http://localhost';
export type SUPERSET_API_BASE = typeof SUPERSET_API_BASE;

/** The Superset API endpoint */
export const SUPERSET_API_ENDPOINT = process.env.REACT_APP_SUPERSET_API_ENDPOINT || 'slice';
export type SUPERSET_API_ENDPOINT = typeof SUPERSET_API_ENDPOINT;

export const SUPERSET_PLANS_SLICE = process.env.REACT_APP_SUPERSET_PLANS_SLICE || '0';
export type SUPERSET_PLANS_SLICE = typeof SUPERSET_PLANS_SLICE;

export const SUPERSET_GOALS_SLICE = process.env.REACT_APP_SUPERSET_GOALS_SLICE || '0';
export type SUPERSET_GOALS_SLICE = typeof SUPERSET_GOALS_SLICE;

export const SUPERSET_JURISDICTIONS_SLICE =
  process.env.REACT_APP_SUPERSET_JURISDICTIONS_SLICE || '0';
export type SUPERSET_JURISDICTIONS_SLICE = typeof SUPERSET_JURISDICTIONS_SLICE;

export const SUPERSET_STRUCTURES_SLICE = process.env.REACT_APP_SUPERSET_STRUCTURES_SLICE || '0';
export type SUPERSET_STRUCTURES_SLICE = typeof SUPERSET_STRUCTURES_SLICE;

/** OpenSRP oAuth2 settings */
export const ENABLE_OPENSRP_OAUTH = process.env.REACT_APP_ENABLE_OPENSRP_OAUTH === 'true';
export type ENABLE_OPENSRP_OAUTH = typeof ENABLE_OPENSRP_OAUTH;
export const OPENSRP_CLIENT_ID = process.env.REACT_APP_OPENSRP_CLIENT_ID || '';
export type OPENSRP_CLIENT_ID = typeof OPENSRP_CLIENT_ID;

// notice the ending is NOT / here
export const OPENSRP_ACCESS_TOKEN_URL =
  process.env.REACT_APP_OPENSRP_ACCESS_TOKEN_URL ||
  'https://reveal-stage.smartregister.org/opensrp/oauth/token';
export type OPENSRP_ACCESS_TOKEN_URL = typeof OPENSRP_ACCESS_TOKEN_URL;

// notice the ending is NOT / here
export const OPENSRP_AUTHORIZATION_URL =
  process.env.REACT_APP_OPENSRP_AUTHORIZATION_URL ||
  'https://reveal-stage.smartregister.org/opensrp/oauth/authorize';
export type OPENSRP_AUTHORIZATION_URL = typeof OPENSRP_AUTHORIZATION_URL;

export const OPENSRP_USER_URL =
  process.env.REACT_APP_OPENSRP_USER_URL ||
  'https://reveal-stage.smartregister.org/opensrp/user-details';
export type OPENSRP_USER_URL = typeof OPENSRP_USER_URL;

export const OPENSRP_OAUTH_STATE = process.env.REACT_APP_OPENSRP_OAUTH_STATE || 'opensrp';
export type OPENSRP_OAUTH_STATE = typeof OPENSRP_OAUTH_STATE;

/** Onadata oAuth2 settings */
export const ENABLE_ONADATA_OAUTH = process.env.REACT_APP_ENABLE_ONADATA_OAUTH === 'true';
export type ENABLE_ONADATA_OAUTH = typeof ENABLE_ONADATA_OAUTH;
export const ONADATA_CLIENT_ID = process.env.REACT_APP_ONADATA_CLIENT_ID || '';
export type ONADATA_CLIENT_ID = typeof ONADATA_CLIENT_ID;

// notice the ending / here
export const ONADATA_ACCESS_TOKEN_URL =
  process.env.REACT_APP_ONADATA_ACCESS_TOKEN_URL || 'https://stage-api.ona.io/o/token/';
export type ONADATA_ACCESS_TOKEN_URL = typeof ONADATA_ACCESS_TOKEN_URL;

// notice the ending / here
export const ONADATA_AUTHORIZATION_URL =
  process.env.REACT_APP_ONADATA_AUTHORIZATION_URL || 'https://stage-api.ona.io/o/authorize/';
export type ONADATA_AUTHORIZATION_URL = typeof ONADATA_AUTHORIZATION_URL;

export const ONADATA_USER_URL =
  process.env.REACT_APP_ONADATA_USER_URL || 'https://stage-api.ona.io/api/v1/user.json';
export type ONADATA_USER_URL = typeof ONADATA_USER_URL;

export const ONADATA_OAUTH_STATE = process.env.REACT_APP_ONADATA_OAUTH_STATE || 'onadata';
export type ONADATA_OAUTH_STATE = typeof ONADATA_OAUTH_STATE;

export const GISIDA_MAPBOX_TOKEN = process.env.REACT_APP_GISIDA_MAPBOX_TOKEN || '';
export type GISIDA_MAPBOX_TOKEN = typeof GISIDA_MAPBOX_TOKEN;

export const GISIDA_ONADATA_API_TOKEN = process.env.REACT_APP_GISIDA_ONADATA_API_TOKEN || '';
export type GISIDA_ONADATA_API_TOKEN = typeof GISIDA_ONADATA_API_TOKEN;
