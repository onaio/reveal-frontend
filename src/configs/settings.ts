/** This is the main configuration module */
import { Providers } from '@onaio/gatekeeper';
import {
  DOMAIN_NAME,
  ENABLE_ONADATA_OAUTH,
  ENABLE_OPENSRP_OAUTH,
  ONADATA_ACCESS_TOKEN_URL,
  ONADATA_AUTHORIZATION_URL,
  ONADATA_CLIENT_ID,
  ONADATA_OAUTH_STATE,
  ONADATA_USER_URL,
  OPENSRP_ACCESS_TOKEN_URL,
  OPENSRP_AUTHORIZATION_URL,
  OPENSRP_CLIENT_ID,
  OPENSRP_OAUTH_STATE,
  OPENSRP_USER_URL,
} from './env';

/** Interfaces and Types */

/** Interface for a Focus Investigation Classification */
export interface Classification {
  code: string;
  name: string;
  description: string;
}

/** Interface that describes location items */
export interface LocationItem {
  identifier: string /** Should match the name of the column in data */;
  level: number /** The HDX-compliant level of the location in the hierarchy */;
  name: string /** The name of the location */;
}

/** Authentication Configs */
const providers: Providers = {};

if (ENABLE_OPENSRP_OAUTH) {
  providers.OpenSRP = {
    accessTokenUri: OPENSRP_ACCESS_TOKEN_URL,
    authorizationUri: OPENSRP_AUTHORIZATION_URL,
    clientId: OPENSRP_CLIENT_ID,
    redirectUri: `${DOMAIN_NAME}/oauth/callback/OpenSRP/`,
    scopes: ['read', 'write'],
    state: OPENSRP_OAUTH_STATE,
    userUri: OPENSRP_USER_URL,
  };
}

if (ENABLE_ONADATA_OAUTH) {
  providers.Ona = {
    accessTokenUri: ONADATA_ACCESS_TOKEN_URL,
    authorizationUri: ONADATA_AUTHORIZATION_URL,
    clientId: ONADATA_CLIENT_ID,
    redirectUri: `${DOMAIN_NAME}/oauth/callback/Ona/`,
    scopes: ['read', 'write'],
    state: ONADATA_OAUTH_STATE,
    userUri: ONADATA_USER_URL,
  };
}

export { providers };

/** Location configs */

/** Location item hierarchy
 * This is a list of locations.  The "level" field will be used to sort the
 * locations hierarchically, from lowest to highest.
 */
export const locationHierarchy: LocationItem[] = [
  {
    identifier: 'province',
    level: 1,
    name: 'Province',
  },
  {
    identifier: 'district',
    level: 2,
    name: 'District',
  },
  {
    identifier: 'canton',
    level: 3,
    name: 'Canton',
  },
  {
    identifier: 'village',
    level: 4,
    name: 'Village',
  },
];

/** Focus investigation configs */

/** Focus Investigation case classifications */
export const FIClassifications: Classification[] = [
  {
    code: 'A1',
    description: 'Indigenous case recorded within the last year.',
    name: 'Active',
  },
  {
    code: 'A2',
    description: 'No indigenous case during the last year, but withing the last 3 years.',
    name: 'Residual Non-Active',
  },
  {
    code: 'B1',
    description: 'Receptive area but no indigenous cases within the last 3 years.',
    name: 'Cleared Receptive',
  },
  {
    code: 'B2',
    description: 'Non-receptive area.',
    name: 'Cleared Non-Receptive',
  },
];

/** Indicators configs */

// thresholds
export const GREEN_THRESHOLD = 0.9;
export type GREEN_THRESHOLD = typeof GREEN_THRESHOLD;
export const YELLOW_THRESHOLD = 0.2;
export type YELLOW_THRESHOLD = typeof YELLOW_THRESHOLD;
export const ORANGE_THRESHOLD = 0.8;
export type ORANGE_THRESHOLD = typeof ORANGE_THRESHOLD;

// 1-3-7 thresholds
export const ONE = 0;
export const ZERO = 0;

export const pointLayerConfig = {
  id: 'single-jurisdiction-20191910',
  layout: {
    'icon-image': 'circle-15',
    'icon-size': 2,
  },
  minzoom: 1,
  paint: {
    'text-color': '#000',
    'text-halo-blur': 1,
    'text-halo-color': '#fff',
    'text-halo-width': 1.3,
  },
  source: {
    data: {
      data: {
        coordinates: [101.177725195885, 15.0658221308165],
        type: 'Point',
      },
      type: 'stringified-geojson',
    },
    minzoom: 1,
    type: 'geojson',
  },
  type: 'symbol',
  visible: false,
};

export const lineLayerConfig = {
  id: 'single-jurisdiction',
  paint: {
    'line-color': '#FFDC00',
    'line-opacity': 1,
    'line-width': 3,
  },
  source: {
    data: {
      data: {
        coordinates: [101.177725195885, 15.0658221308165],
        type: 'Point',
      },
      type: 'stringified-geojson',
    },
    type: 'geojson',
  },
  type: 'line',
  visible: false,
};

export const fillLayerConfig = {
  id: 'single-jurisdiction',
  paint: {
    'fill-color': '#FFDC00',
    'fill-opacity': 1,
    'fill-outline-color': '#FFDC00',
  },
  source: {
    data: {
      data: {
        coordinates: [101.177725195885, 15.0658221308165],
        type: 'Point',
      },
      type: 'stringified-geojson',
    },
    type: 'geojson',
  },
  type: 'fill',
  visible: false,
};
