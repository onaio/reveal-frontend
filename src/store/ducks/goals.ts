import { Dictionary } from '@onaio/utils';
import { get, keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { PlanActionCodesType } from '../../components/forms/PlanForm/types';

/** the reducer name */
export const reducerName = 'goals';

/** interface for goal Object */
export interface Goal {
  action_code: PlanActionCodesType;
  action_description: string;
  action_prefix: string;
  action_reason: string;
  action_title: string;
  completed_task_count: number;
  goal_comparator: string;
  goal_id: string;
  goal_unit: string;
  goal_value: number;
  id: string;
  jurisdiction_id: string;
  measure: string;
  plan_id: string;
  task_business_status_map: Dictionary;
  task_count: number;
}
// actions
/** GOALS_FETCHED action type */
export const GOALS_FETCHED = 'reveal/reducer/goals/GOALS_FETCHED';

/** SET_CURRENT_GOAL action type */
export const SET_CURRENT_GOAL = 'reveal/reducer/goals/SET_CURRENT_GOAL';

/** REMOVE_GOALS action type */
export const REMOVE_GOALS = 'reveal/reducer/goals/REMOVE_GOALS';

/** interface for authorize action */
interface FetchGoalsAction extends AnyAction {
  goalsById: { [key: string]: Goal };
  type: typeof GOALS_FETCHED;
}

/** interface for setting current goal action */
interface SetCurrentGoalAction extends AnyAction {
  currentGoal: string | null;
  type: typeof SET_CURRENT_GOAL;
}

/** Interface for remove goals action */
interface RemoveGoalsAction extends AnyAction {
  type: typeof REMOVE_GOALS;
  goalsById: {};
}

/** Create type for Goal reducer action */
export type GoalActionTypes =
  | FetchGoalsAction
  | SetCurrentGoalAction
  | RemoveGoalsAction
  | AnyAction;

/** interface for Goal state */
interface GoalState {
  goalsById: { [key: string]: Goal[] } | {};
  currentGoal: string | null;
}

/** immutable Goal state */
export type ImmutableGoalState = GoalState & SeamlessImmutable.ImmutableObject<GoalState>;

/** initial Goal state */
const initialState: ImmutableGoalState = SeamlessImmutable({
  currentGoal: null,
  goalsById: {},
});

/** the Goal reducer function */
export default function reducer(state = initialState, action: GoalActionTypes): ImmutableGoalState {
  switch (action.type) {
    case GOALS_FETCHED:
      return SeamlessImmutable({
        ...state,
        goalsById: { ...state.goalsById, ...action.goalsById },
      });
    case SET_CURRENT_GOAL:
      return SeamlessImmutable({
        ...state,
        currentGoal: action.currentGoal,
      });
    case REMOVE_GOALS:
      return SeamlessImmutable({
        ...state,
        goalsById: action.goalsById,
      });
    default:
      return state;
  }
}

// action creators

/** fetch Goals creator
 * @param {Goal[]} goalsList - array of goal objects
 * @returns {FetchGoalsAction} FetchGoalsAction
 */
export const fetchGoals = (goalsList: Goal[] = []): FetchGoalsAction => {
  return {
    goalsById: keyBy(
      goalsList.map((item: Goal) => {
        if (typeof item.task_business_status_map === 'string') {
          item.task_business_status_map = JSON.parse(item.task_business_status_map);
        }
        return item;
      }),
      goal => goal.id
    ),
    type: GOALS_FETCHED,
  };
};

/** fetch Current Goal
 * @param currentGoal - current selected goal type string if none defaults to null
 * @returns {SetchCurrentGoalAction} setchCurrentGoalAction
 */
export const setCurrentGoal = (currentGoal: string | null): SetCurrentGoalAction => {
  return {
    currentGoal,
    type: SET_CURRENT_GOAL,
  };
};

// Actions

export const removeGoalsAction: RemoveGoalsAction = {
  goalsById: {},
  type: REMOVE_GOALS,
};

// selectors

/** get goals by id
 * @param {Partial<Store>} state - the redux store
 */
export function getGoalsById(state: Partial<Store>): { [key: string]: Goal } {
  return (state as any)[reducerName].goalsById;
}

/** get one goal using its id
 * @param {Partial<Store>} state - the redux store
 * @param {string} id - the goal id
 * @returns {Goal|null} a goal or null
 */
export function getGoalById(state: Partial<Store>, id: string): Goal | null {
  return get((state as any)[reducerName].goalsById, id) || null;
}

/** get goals by plan id
 * @param {Partial<Store>} state - the redux store
 * @param {string} planId - the plan id
 * @returns {Goal[]} an array of goals
 */
export function getGoalsByPlanId(state: Partial<Store>, planId: string): Goal[] {
  return values((state as any)[reducerName].goalsById).filter((e: Goal) => e.plan_id === planId);
}

/** get goals by goal id
 * @param {Partial<Store>} state - the redux store
 * @param {string} goalId - the goal id
 * @returns {Goal[]} an array of goals
 */
export function getGoalsByGoalId(state: Partial<Store>, goalId: string): Goal[] {
  return values((state as any)[reducerName].goalsById).filter((e: Goal) => e.goal_id === goalId);
}

/** get goals by jurisdiction id
 * @param {Partial<Store>} state - the redux store
 * @param {string} jurisdictionId - the jurisdiction id
 * @returns {Goal[]} an array of goals
 */
export function getGoalsByJurisdictionId(state: Partial<Store>, jurisdictionId: string): Goal[] {
  return values((state as any)[reducerName].goalsById).filter(
    (e: Goal) => e.jurisdiction_id === jurisdictionId
  );
}

/** get goals by plan id and jurisdiction id
 * @param {Partial<Store>} state - the redux store
 * @param {string} planId - the plan id
 * @param {string} jurisdictionId - the jurisdiction id
 * @returns {Goal[]} an array of goals
 */
export function getGoalsByPlanAndJurisdiction(
  state: Partial<Store>,
  planId: string,
  jurisdictionId: string
): Goal[] {
  return values((state as any)[reducerName].goalsById).filter(
    (e: Goal) => e.plan_id === planId && e.jurisdiction_id === jurisdictionId
  );
}
/** get currently selected goal */
export function getCurrentGoal(state: Partial<Store>) {
  return (state as any)[reducerName].currentGoal;
}
