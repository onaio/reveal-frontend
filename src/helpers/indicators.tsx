import ElementMap from '@onaio/element-map';
import { Dictionary } from '@onaio/utils';
import { keys } from 'lodash';
import * as React from 'react';
import { CellInfo } from 'react-table';
import { GREEN, ORANGE, RED, WHITE, YELLOW } from '../colors';
import { PERSONS, STRUCTURES } from '../configs/lang';
import {
  GREEN_THRESHOLD,
  IndicatorThresholds,
  indicatorThresholdsIRS,
  irsReportingCongif,
  ORANGE_THRESHOLD,
  YELLOW_THRESHOLD,
  ZERO,
} from '../configs/settings';
import { BLOOD_SCREENING_CODE, CASE_CONFIRMATION_CODE } from '../constants';
import { IndicatorThresholdItemPercentage, roundToPrecision } from '../helpers/utils';
import { Goal } from '../store/ducks/goals';

/** Enum describing operators */
enum Operators {
  Equal = '=',
  GreaterThan = '>',
  GreaterThanOrEqual = '>=',
  LessThan = '<',
  LessThanOrEqual = '<=',
}

/** Get the level of achievement towards the goal target as a ratio
 * @param {Goal} goal - the goal
 * @returns {number} ratio Achieved
 */
export function goalRatioAchieved(goal: Goal): number {
  let ratioAchieved: number = 0;
  const achievedValue: number = goal.completed_task_count;
  const totalAttempts: number = goal.task_count;
  let targetValue: number = goal.goal_value;

  // set the actual target value for percentages
  if (goal.goal_unit.toLowerCase() === 'percent') {
    targetValue = roundToPrecision((targetValue * totalAttempts) / 100);
  }

  if (targetValue === 0) {
    return 0; /** Not yet supported */
  }

  // get the completed ratio
  ratioAchieved = roundToPrecision(achievedValue / targetValue, 2);

  if (
    goal.goal_comparator === Operators.LessThan ||
    goal.goal_comparator === Operators.LessThanOrEqual
  ) {
    // in this case we are targeting a reduction
    ratioAchieved = 0; /** Not yet supported */
  }

  return ratioAchieved;
}

/** interface for Goal report */
export interface GoalReport {
  achievedValue: number /** the achieved value */;
  percentAchieved: number /** progress towards goal achievement */;
  prettyPercentAchieved: string | null /** pretty string of percentAchieved */;
  targetValue: number /** the target value */;
  goalUnit: string /** goal_unit */;
}

/** Utility function to get an object containing values for goal indicators
 * @param {Goal} goal - the goal
 * @returns {GoalReport} the Goal Report object
 */
export function getGoalReport(goal: Goal): GoalReport {
  const percentAchieved = goalRatioAchieved(goal);
  let targetValue;
  let goalUnit: string = goal.goal_unit;
  if (goal.goal_unit.toLowerCase() !== 'percent') {
    targetValue = goal.goal_value;
  } else {
    goalUnit =
      [CASE_CONFIRMATION_CODE as string, BLOOD_SCREENING_CODE as string].indexOf(goal.action_code) >
      -1
        ? PERSONS
        : STRUCTURES;
    targetValue = Math.round((goal.goal_value * goal.task_count) / 100);

    /** Use goal.goal_value as target if unit==='Percent' && goal.task_count === 0 */
    targetValue = targetValue === 0 ? goal.goal_value : targetValue;
  }

  return {
    achievedValue: goal.completed_task_count,
    goalUnit,
    percentAchieved,
    prettyPercentAchieved: IndicatorThresholdItemPercentage(percentAchieved),
    targetValue,
  };
}

/** Returns a table cell rendered with different colors based on focus
 * investigation response adherence conditional formatting
 */
export function getFIAdherenceIndicator(cell: CellInfo) {
  return (
    <div
      className="indicator-container"
      style={{
        backgroundColor:
          cell.value >= GREEN_THRESHOLD
            ? GREEN
            : cell.value >= ORANGE_THRESHOLD
            ? ORANGE
            : cell.value >= YELLOW_THRESHOLD
            ? RED
            : YELLOW,
      }}
    >
      {IndicatorThresholdItemPercentage(cell.value)}
    </div>
  );
}

/** Returns a value formatted for Focus Investigation 1-3-7 adherence
 */
export function get137Value(value: number): string {
  if (value < 0) {
    return '<1';
  }
  return `${value}`;
}

/** Returns a table cell rendered with different colors based on focus
 * investigation 1-3-7 adherence conditional formatting
 */
export function get137AdherenceIndicator(cell: CellInfo) {
  return (
    <div
      className="137-container"
      style={{
        color: cell.value > ZERO ? GREEN : cell.value < ZERO ? RED : ORANGE,
      }}
    >
      {cell.value}d to go
    </div>
  );
}

/** Renders a row of Focus Investigation classifications */
export function renderClassificationRow(rowObject: Dictionary) {
  return (
    <tr key={rowObject.code} className="definitions">
      <ElementMap items={[rowObject.code, rowObject.name, rowObject.description]} HTMLTag="td" />
    </tr>
  );
}

/** Determines the color of a indicator based on cell.value and threshold configs
 * @param {CellInfo} cell - the ReactTable.Cell being rendered in an indicator drill-down table
 * @param {IndicatorThresholds} thresholds - the threshold configuration from settings
 * @param {boolean} doMakeDecimal -  determines if the cell value needs to be divided by 100
 * @returns {string} - the value corresponding the cell.value according to the reporting config
 */
export function getThresholdColor(
  cell: CellInfo,
  thresholds: IndicatorThresholds,
  doMakeDecimal: boolean = false
) {
  // get the keys of the thresholds and sort them by value
  const thresholdKeys = keys(thresholds).sort((a, b) => thresholds[a].value - thresholds[b].value);
  const value = doMakeDecimal ? cell.value / 100 : cell.value;

  // loop through thresholds and evaluate value against cell.value
  let k: string = '';
  for (k of thresholdKeys) {
    if (thresholds[k].orEquals ? value <= thresholds[k].value : value < thresholds[k].value) {
      return thresholds[k].color;
    }
  }

  // if value is greater than the biggest threshold, assign it the biggest threshold's color
  if (k.length) {
    return thresholds[k].color;
  }

  // fallback to white
  return WHITE;
}

/** Renders an indicator Cell based on cell.value and threshold configs
 * @param {CellInfo} cell - the ReactTable.Cell being rendered in an indicator drilldown table
 * @param {string} configId - the key vause used to get the custom reporting configs
 * @returns {React.ReactElement} - the ReactTable.Cell element to be rendered for the indicator
 */
export function getThresholdAdherenceIndicator(cell: CellInfo, configId: string) {
  // determine if cell.value is a number
  const isNumber = !Number.isNaN(Number(cell.value));
  // get thresholds config from settings
  const thresholds: IndicatorThresholds | null =
    isNumber && irsReportingCongif[configId]
      ? irsReportingCongif[configId].indicatorThresholds
      : null;
  // determine cell background color
  const cellColor = thresholds ? getThresholdColor(cell, thresholds) : WHITE;

  return (
    <div className="irs-report-indicator-container" style={{ backgroundColor: cellColor }}>
      {isNumber ? IndicatorThresholdItemPercentage(cell.value) : 'NaN'}
    </div>
  );
}

/** Renders an indicator Cell based on cell.value and threshold configs
 * @param {CellInfo} cell - the ReactTable.Cell being rendered in an indicator drilldown table
 * @param {IndicatorThresholds | null} thresholds - the indicator thresholds
 * @returns {React.ReactElement} - the ReactTable.Cell element to be rendered for the indicator
 */
export function getIRSThresholdAdherenceIndicator(
  cell: CellInfo,
  thresholds: IndicatorThresholds | null = indicatorThresholdsIRS
) {
  // determine if cell.value is a number
  const isNumber = !Number.isNaN(Number(cell.value));
  // determine cell background color
  const cellColor = thresholds ? getThresholdColor(cell, thresholds) : WHITE;

  return (
    <div className="irs-report-indicator-container" style={{ backgroundColor: cellColor }}>
      {isNumber ? IndicatorThresholdItemPercentage(cell.value) : 'NaN'}
    </div>
  );
}
