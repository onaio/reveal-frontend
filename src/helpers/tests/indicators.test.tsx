import { map } from 'lodash';
import { PERSONS, STRUCTURES } from '../../configs/lang';
import { Goal } from '../../store/ducks/goals';
import * as fixtures from '../../store/ducks/tests/fixtures';
import { getGoalReport, goalRatioAchieved } from '../indicators';

describe('helpers/indicators', () => {
  it('goalPercentAchieved works', () => {
    // returns correct percentage for all goal_unit values
    const expected: number[] = [0, 0, 0, 1, 0.27, 0.36, 0.03];
    const got = map(fixtures.goals, goal => goalRatioAchieved(goal as Goal));
    expect(got).toEqual(expected);
  });

  it('goalPercentAchieved works for reducing goals', () => {
    const goal = {
      action_code: 'Blood Screening',
      action_description:
        'Visit all residential structures (100%) within a 1 km radius of a confirmed index case and test each registered person',
      action_prefix: '4',
      action_reason: 'Investigation',
      action_title: 'RACD Blood screening',
      completed_task_count: 100,
      goal_comparator: '<',
      goal_id: 'RACD_blood_screening_1km_radius',
      goal_unit: 'pints',
      goal_value: 30,
      id: '19b86421-3cb2-5698-9f11-c1bdafbe5e6d',
      jurisdiction_id: '1337',
      measure: 'Percent of registered people tested',
      plan_id: '1337',
      task_business_status_map: {},
      task_count: 100,
    };
    expect(goalRatioAchieved(goal as Goal)).toEqual(0);
  });

  it('goalPercentAchieved works for target that is 0', () => {
    const goal = {
      action_code: 'Blood Screening',
      action_description:
        'Visit all residential structures (100%) within a 1 km radius of a confirmed index case and test each registered person',
      action_prefix: '4',
      action_reason: 'Investigation',
      action_title: 'RACD Blood screening',
      completed_task_count: 100,
      goal_comparator: '<',
      goal_id: 'RACD_blood_screening_1km_radius',
      goal_unit: 'percent',
      goal_value: 0,
      id: '19b86421-3cb2-5698-9f11-c1bdafbe5e6d',
      jurisdiction_id: '1337',
      measure: 'Percent of registered people tested',
      plan_id: '1337',
      task_business_status_map: {},
      task_count: 100,
    };
    expect(goalRatioAchieved(goal as Goal)).toEqual(0);
  });

  it('getGoalReport works', () => {
    expect(getGoalReport(fixtures.goal1 as Goal)).toEqual({
      achievedValue: 0,
      goalUnit: 'each',
      percentAchieved: 0,
      prettyPercentAchieved: '0%',
      targetValue: 1,
    });
    expect(getGoalReport(fixtures.goal2 as Goal)).toEqual({
      achievedValue: 0,
      goalUnit: 'each',
      percentAchieved: 0,
      prettyPercentAchieved: '0%',
      targetValue: 3,
    });
    expect(getGoalReport(fixtures.goal3 as Goal)).toEqual({
      achievedValue: 0,
      goalUnit: PERSONS,
      percentAchieved: 0,
      prettyPercentAchieved: '0%',
      targetValue: 100,
    });
    expect(getGoalReport(fixtures.goal4 as Goal)).toEqual({
      achievedValue: 3,
      goalUnit: 'traps',
      percentAchieved: 1,
      prettyPercentAchieved: '100%',
      targetValue: 3,
    });
    expect(getGoalReport(fixtures.goal5 as Goal)).toEqual({
      achievedValue: 4,
      goalUnit: STRUCTURES,
      percentAchieved: 0.27,
      prettyPercentAchieved: '27%',
      targetValue: 15,
    });
    expect(getGoalReport(fixtures.goal6 as Goal)).toEqual({
      achievedValue: 4,
      goalUnit: STRUCTURES,
      percentAchieved: 0.36,
      prettyPercentAchieved: '36%',
      targetValue: 11,
    });
    expect(getGoalReport(fixtures.goal7 as Goal)).toEqual({
      achievedValue: 3,
      goalUnit: STRUCTURES,
      percentAchieved: 0.6,
      prettyPercentAchieved: '60%',
      targetValue: 5,
    });
  });
});
