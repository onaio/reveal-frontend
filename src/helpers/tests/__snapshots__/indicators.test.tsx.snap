// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers/indicators getIRSLiteThresholdAdherenceIndicator works: 0.11 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "#dddddd",
    }
  }
>
  11%
</div>
`;

exports[`helpers/indicators getIRSLiteThresholdAdherenceIndicator works: 0.21 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "#FF4136",
    }
  }
>
  21%
</div>
`;

exports[`helpers/indicators getIRSLiteThresholdAdherenceIndicator works: 0.81 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "#FFDC00",
    }
  }
>
  81%
</div>
`;

exports[`helpers/indicators getIRSLiteThresholdAdherenceIndicator works: 0.91 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "#2ECC40",
    }
  }
>
  91%
</div>
`;

exports[`helpers/indicators getIRSLiteThresholdAdherenceIndicator works: invalid number 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "inherit",
    }
  }
>
  
</div>
`;

exports[`helpers/indicators getIRSThresholdAdherenceIndicator works: 0.11 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "#dddddd",
    }
  }
>
  11%
</div>
`;

exports[`helpers/indicators getIRSThresholdAdherenceIndicator works: 0.21 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "#FF4136",
    }
  }
>
  21%
</div>
`;

exports[`helpers/indicators getIRSThresholdAdherenceIndicator works: 0.81 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "#FFDC00",
    }
  }
>
  81%
</div>
`;

exports[`helpers/indicators getIRSThresholdAdherenceIndicator works: 0.91 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "#2ECC40",
    }
  }
>
  91%
</div>
`;

exports[`helpers/indicators getIRSThresholdAdherenceIndicator works: invalid number 1`] = `
<div
  className="irs-report-indicator-container"
  style={
    Object {
      "backgroundColor": "inherit",
    }
  }
>
  
</div>
`;
